driver-core/src/main/com/mongodb/operation/AsyncChangeStreamBatchCursor.java

```diff
@@ -167,16 +168,25 @@ public void onResult(final List<RawBsonDocument> result, final Throwable t) {
     }
 
     private void retryOperation(final AsyncBlock asyncBlock, final SingleResultCallback<List<RawBsonDocument>> callback) {
-        changeStreamOperation.setChangeStreamOptionsForResume(resumeToken);
-        changeStreamOperation.executeAsync(binding, new SingleResultCallback<AsyncBatchCursor<T>>() {
+        binding.getReadConnectionSource(new SingleResultCallback<AsyncConnectionSource>() {
             @Override
-            public void onResult(final AsyncBatchCursor<T> result, final Throwable t) {
+            public void onResult(final AsyncConnectionSource result, final Throwable t) {
                 if (t != null) {
                     callback.onResult(null, t);
                 } else {
-                    wrapped = ((AsyncChangeStreamBatchCursor<T>) result).getWrapped();
-                    binding.release(); // release the new change stream batch cursor's reference to the binding
-                    resumeableOperation(asyncBlock, callback);
+                    changeStreamOperation.setChangeStreamOptionsForResume(resumeToken, result.getServerDescription().getMaxWireVersion());
```

Comment:
Is it an issue that we get one source here and set the options and then get another potentially different source when executing?  I'm not sure I can think of a way around it and if there is mixed cluster then I imagine the behavior is undefined.

Reply:
I don't see a way around this. If we are resuming an operation, we need to make sure we have a valid read connection source.

Reply:
Agreed, I can't see a clean way to do it all with a single connection source.